src/Appwrite/Platform/Tasks/Backup.php

```diff
@@ -0,0 +1,318 @@
+
+        $args = [
+            'xtrabackup',
+            '--user=' . $this->dsn->getUser(),
+            '--password=' . $this->dsn->getPassword(),
+            '--host=' . $this->dsn->getHost(),
+            '--port=' . $this->dsn->getPort(),
+            '--backup',
+            '--strict',
+            '--history="' . $this->database . '|' . pathinfo($filename, PATHINFO_FILENAME) . '"', // PERCONA_SCHEMA.xtrabackup_history
+            '--slave-info',
+            '--safe-slave-backup',
+            '--safe-slave-backup-timeout=300',
+            '--check-privileges', // checks if Percona XtraBackup has all the required privileges.
+            '--target-dir=' . $target,
+            '--compress=' . self::COMPRESS_ALGORITHM,
+            '--compress-threads=' . $this->processors, // Better not using all processors
+            '--parallel=' . $this->processors,
+            '--rsync', // https://docs.percona.com/percona-xtrabackup/8.0/accelerate-backup-process.html
+            '2> ' . $logfile,
+        ];
+
+        $cmd = 'docker exec ' . $this->xtrabackupContainerId . ' ' . implode(' ', $args);
+        shell_exec($cmd);
+
+        $stderr = shell_exec('tail -1 ' . $logfile);
+        if (!str_contains($stderr, 'completed OK!')) {
+            shell_exec('rm -rf ' . $target);
+            throw new Exception(' Backup failed: ' . $stderr);
+        }
+
+        if (!unlink($logfile)) {
+            throw new Exception('Error deleting: ' . $logfile);
+        }
+
+        self::log('Xtrabackup end ' . (microtime(true) - $start) . ' seconds');
+    }
+
+    /**
+     * @throws Exception
+     */
+    public function tar(string $directory, string $file)
+    {
+        $start = microtime(true);
+        self::log('Tar start');
+
+        $stdout = '';
+        $stderr = '';
+        $cmd = 'cd ' . $directory . ' && tar zcf ' . $file . ' . && cd ' . getcwd();
+        Console::execute($cmd, '', $stdout, $stderr);
+        if (!empty($stderr)) {
+            throw new Exception('Tar failed:' . $stderr);
+        }
+
+        if (!file_exists($file)) {
+            throw new Exception('Can\'t find tar file: ' . $file);
+        }
+
+        $filesize = \filesize($file);
+        if ($filesize < 5 * 1024 * 1024) {
+            throw new Exception('Tar file size is very small: ' . $file);
+        }
+
+        self::log('Tar took ' . (microtime(true) - $start) . ' seconds');
+    }
+
+    /**
+     * @throws Exception
+     */
+    public function upload(string $file, Device $local)
+    {
+        $start = microtime(true);
+        self::log('Upload start');
+        $filename = basename($file);
+
+        if (!$this->s3->exists('/')) {
+            throw new Exception('Can\'t read s3 root directory');
+        }
+
+        $destination = $this->s3->getRoot() . '/' . $filename;
+
+        try {
+            if (!$local->transfer($file, $destination, $this->s3)) {
+                throw new Exception('Error uploading to ' . $destination);
```

Comment:
Let's add a retry loop here with 5 max attempts before throwing

Reply:
Done! 
Added CONSTS for retry options.